sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -93,6 +93,14 @@ object TypeCoercion {
       Some(t2)
     case (t1: DecimalType, t2: IntegralType) if t1.isWiderThan(t2) =>
       Some(t1)
+    case (t1: DecimalType, t2: DecimalType) =>
+      // Handle two decimal type here and don't allow precision loss.
+      val widerType = DecimalPrecision.widerDecimalType(t1, t2)
+      if (widerType.isWiderThan(t1) && widerType.isWiderThan(t2)) {
+        Some(widerType)
+      } else {
+        None
+      }
```

Comment:
This code looks suspicious... I personally think this issue should be fixed only in `InConversion` instead of `findTightestCommonType`. That's because the change of `findTightestCommonType` can affect type coercion in the other operations... cc: @mgaido91 @cloud-fan

Reply:
But I think for two decimalType, such as Decimal(3,0) and Decimal(3,2), their tightest common type should be Decimal(5,2).
It is consistent with the method name findTightestCommonType.

Reply:
is this related to the specific bug? If not let's open another PR to do it.

Reply:
It's more reasonable to fix `InConversion`. I think it's wrong that `In` and `InSubquery` have different type coercion logic.

Reply:
Thanks @cloud-fan  I will deep check it.

Reply:
Actually `InConversion` and `BinaryComparison` also have different type coercion logic: https://github.com/apache/spark/pull/22038

Reply:
I find a smilar implementation:
https://github.com/apache/spark/blob/e46e487b0831b39afa12ef9cff9b9133f111921b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala#L316-L324

Reply:
> I think it's wrong that In and InSubquery have different type coercion logic.

I agree on this. Please see https://github.com/apache/spark/pull/19635, where I tried to fix this....